File: year2/advanced_programming/Lab7/lab7/Repository/InMemoryRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using lab7.Domain;
using lab7.Domain.Validator;

namespace lab7.Repository
{
    class InMemoryRepository<ID, E> : Repository<ID, E> where E : Entity<ID>
    {
        protected Validator<E> vali;

        protected IDictionary<ID, E> entities = new Dictionary<ID, E>();

        public InMemoryRepository(Validator<E> vali)
        {
            this.vali = vali;
        }

        public E Delete(ID id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<E> FindAll()
        {
            return entities.Values.ToList<E>();
        }

        public E FindOne(ID id)
        {
            throw new NotImplementedException();
        }

        public E Save(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("entity must not be null");
            this.vali.Validate(entity);
            if (this.entities.ContainsKey(entity.ID))
            {
                return entity;
            }
            this.entities[entity.ID] = entity;
            return default(E);
        }

        public E Update(E entity)
        {
            throw new NotImplementedException();
        }
    }
}

File: year2/advanced_programming/Lab3/src/main/java/socialnetwork/repository/memory/InMemoryRepository.java
package socialnetwork.repository.memory;

import socialnetwork.domain.Entity;
import socialnetwork.domain.Tuple;
import socialnetwork.domain.validators.Validator;
import socialnetwork.repository.Repository;


import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID, E> {

    private Validator<E> validator;
    Map<ID, E> entities;

    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        entities = new HashMap<ID, E>();
    }


    @Override
    public Optional<E> findOne(ID id) {
        if (id == null)
            throw new IllegalArgumentException("id must be not null");
        E e = entities.get(id);
        if (e == null)
            return Optional.empty();
        else
            return Optional.of(e);
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public Optional<E> save(E entity) {
        if (entity == null)
            throw new IllegalArgumentException("entity must be not null");
        validator.validate(entity);
        for (E e : entities.values()) {
            if (e.getId().equals(entity.getId())) {
                return Optional.of(e);
            }
        }
        entities.put(entity.getId(), entity);
        return Optional.empty();
    }

    @Override
    public Optional<E> delete(ID id) {
        if (id == null)
            throw new IllegalArgumentException("id must be not null");
        Map<ID, E> entitiesCopy = entities;
        for (E e : entitiesCopy.values()) {
            if (e.getId().equals(id)) {
                entities.remove(e.getId());
                return Optional.of(e);
            }
        }
        return Optional.empty();
    }

    @Override
    public Optional<E> update(E entity) {
        if (entity == null)
            throw new IllegalArgumentException("entity must be not null!");
        validator.validate(entity);

        if (entities.get(entity.getId()) != null) {
            entities.put(entity.getId(), entity);
            return Optional.empty();
        }
        return Optional.of(entity);
    }

}
